src/v/cluster/controller.cc

```diff
@@ -246,11 +246,48 @@ ss::future<> controller::start() {
           return _stm.invoke_on(controller_stm_shard, &controller_stm::start);
       })
       .then([this] {
-          return _stm.invoke_on(controller_stm_shard, [](controller_stm& stm) {
-              // we do not have to use timeout in here as all the batches to
-              // apply have to be accesssible
-              return stm.wait(stm.bootstrap_last_applied(), model::no_timeout);
-          });
+          auto disk_commit_offset = _raft0->log().offsets().committed_offset;
+
+          return _stm
+            .invoke_on(
+              controller_stm_shard,
+              [disk_commit_offset](controller_stm& stm) {
+                  // we do not have to use timeout in here as all the batches to
+                  // apply have to be accesssible
+                  auto last_applied = stm.bootstrap_last_applied();
+
+                  // Consistency check: on a bug-free system, the last_applied
+                  // value in the kvstore always points to data on disk.
+                  // However, if there is a bug, or someone has removed some log
+                  // segments out of band, we will hang trying to read up to
+                  // last_applied. Mitigate this by clamping it to the top of
+                  // the log on disk.
+                  if (last_applied > disk_commit_offset) {
```

Comment:
maybe we should check if the log is empty and only then skip waiting for last applied ?

Reply:
The general `>` condition handles cases like:
- start once with empty log
- only partially recover raft log
- then restart again
- the second time we come up, last_applied is still ahead of log, but log is not empty.